src/adr/0036-against-use-of-conditional-statements-in-guice-modules.md

```diff
@@ -0,0 +1,110 @@
+
+James propose several implementations for each of the interfaces it defines. The number of possible combinations of implementations is thus really high (like factorial(n) with n > 10). It makes it unpractical to run tests for each possible interface
+combination. We rather run integration tests for combinations that we decide brings the more value to the users. By overriding any class, Spring defeats 
+this testing logic. Knowing which component implementation combines well with which other one is not intuitive to users,
+hard to document and test.
+
+We thus decided to rely on Guice. Components are defined by code in a static fashion. Overriding components requires 
+explicit code modification and recompilation, thus warning the user about the impact of the choices he does, and lowering 
+the project's responsibility.
+
+With Guice we expose only supported, well tested combinations of components, thus addressing the combination issue.
+
+Instead of having a single big application being able to instantiate each and every component application, we have 
+several products defining their dependencies in a minimalistic way, relying only on the components implementation that 
+are needed.
+
+Here is the list of products we provide:
+
+ - In-Memory: A memory based James server, mainly for testing purposes
+ - Distributed James: A scalable James server, storing data in various data stores. Cassandra is used for metadata, 
+ ElasticSearch for search, RabbitMQ for messaging, and ObjectStorage for blobs.
+ - Cassandra: An implementation step toward Distributed James. It does not include messaging and ObjectStorage and should not be run in a cluster way but is still relevant for good performance.
+ - JPA: A JPA and Lucene based implementation of James. Only Derby driver is currently supported.
+ - JPA with SMTP only using derby: A minimalist SMTP server based on JPA storage technology and Derby driver
+ - JPA with SMTP only using mariadb: A minimalist SMTP server based on JPA storage technology and MariaDB driver
+
+Some components however do have several implementations a user can choose from in a given product. This is the case for:
+
+ - BlobExport: Exporting a blob from the blobStore to an external user. Two implementations are currently supported: 
+ localFiles and LinShare.
+ - Text extraction: Extracting text from attachment to enable attachment search. There is a Tika implementation, but 
+ lighter JSOUP based options are also available.
+
+In order to keep the number of products low, we decided to use conditional statements in modules based on the 
+configuration to select which one to enable at runtime. Eventually defeating the Guice adoption goals mentioned above.
+
+Finally, Blob Storing technology offers a wide combination of technologies:
+
+ - ObjectStorage in itself could implement either Swift APIs or Amazon S3 APIs
+ - We decided to keep supporting Cassandra for blob storing as an upgrade solution from Cassandra product to Distributed 
```

Comment:
do we? is there some documentation about this?

Reply:
I remember it is the main reason.

No. Nothing more than blob.properties

Reply:
> I remember it is the main reason.

What about actually wanting to store data in Cassandra? (For a dataset < 1 TB it makes perfect sense IMO)

Reply:
As metionned by Matthieu:

```
We decided to keep supporting Cassandra for blob storing as an upgrade solution from Cassandra product to Distributed 
James for existing users. This option also makes sense for small data-sets (less than a TB) where storage cost are less 
of an issue and don't need to be taken into account when reasoning about performance.
```